lib/collection/src/collection_manager/segments_updater.rs

```diff
@@ -157,39 +162,106 @@ fn upsert_with_payload(
     Ok(res)
 }
 
+/// Sync points within a given range
+///
+/// 1. Retrieve existing points for a range
+/// 2. Remove points, which are not present in the sync operation
+/// 3. Retrieve overlapping points, detect which one of them are changed
+/// 4. Select new points
+/// 5. Upsert points which differ from the stored ones
+///
+/// Returns:
+///     (number of deleted points, number of new points, number of updated points)
+pub(crate) fn sync_points(
+    segments: &SegmentHolder,
+    op_num: SeqNumberType,
+    from_id: Option<PointIdType>,
+    to_id: Option<PointIdType>,
+    points: &[PointStruct],
+) -> CollectionResult<(usize, usize, usize)> {
+    let id_to_point = points
+        .iter()
+        .map(|p| (p.id, p))
+        .collect::<HashMap<PointIdType, &PointStruct>>();
+    let sync_points: HashSet<_> = points.iter().map(|p| p.id).collect();
+    // 1. Retrieve existing points for a range
+    let stored_point_ids: HashSet<_> = segments
+        .iter()
+        .flat_map(|(_, segment)| segment.get().read().read_range(from_id, to_id))
+        .collect();
+    // 2. Remove points, which do not present in the sync operation
+    let points_to_remove: Vec<_> = stored_point_ids.difference(&sync_points).copied().collect();
+    let deleted = delete_points(segments, op_num, points_to_remove.as_slice())?;
+    // 3. Retrieve overlapping points, detect which one of them are changed
+    let existing_point_ids: Vec<_> = stored_point_ids
+        .intersection(&sync_points)
+        .copied()
+        .collect();
+
+    let mut points_to_update: Vec<_> = Vec::new();
+    let _num_updated = segments.read_points(existing_point_ids.as_slice(), |id, segment| {
+        let vector = segment.vector(id)?;
+        let payload = segment.payload(id)?;
+        let point = id_to_point.get(&id).unwrap();
+        let payload_match = match point.payload {
+            Some(ref p) => p == &payload,
+            None => Payload::default() == payload,
+        };
+
+        if point.vector != vector || !payload_match {
```

Comment:
Does it make sense performance wise to not compute `payload_match` if the `vector` part is identical?
We would spare the computation if the `vector` part has changed.